### ISSUE TITLE
optimizer state is corrupted after torch DCP save/load round-trip

### ISSUE THREAD

**Issue by hemildesai** [databricks/flashoptim#3]
When using `FlashAdamW` with PyTorch FSDP2 (`fully_shard`) and saving/loading optimizer state via `torch.distributed.checkpoint` (DCP), the resumed optimizer produces catastrophically wrong updates. The first forward pass after resume matches exactly (delta=0), but the first `optimizer.step()` causes loss to explode from ~2.4 to ~1785, indicating the restored optimizer state is silently corrupted.

## Environment

- **flashoptim**: 0.1.2
- **PyTorch**: 2.9.0+cu129
- **transformers**: 5.0.0
- **GPU**: 2x NVIDIA GPUs (tested with `torchrun --nproc-per-node 2`)
- **Model**: Qwen/Qwen3-0.6B (pretrained, bf16)
- **Data**: allenai/tulu-3-sft-mixture

### Script

The standalone repro script trains a pretrained Qwen3-0.6B model with `FlashAdamW` for 20 steps on real chat data, saves the optimizer state via DCP, loads it into a fresh optimizer (with identical model weights copied over), and then compares continued training losses between the original and resumed paths.

Save the script below as `test_flashoptim_dcp.py`:

```python
import os
import shutil

import torch
import torch.distributed as dist
import torch.distributed.checkpoint as dcp
from datasets import load_dataset
from torch.distributed.checkpoint.state_dict import (
    StateDictOptions,
    get_optimizer_state_dict,
    set_optimizer_state_dict,
)
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.fsdp import fully_shard
from transformers import AutoModelForCausalLM, AutoTokenizer

from flashoptim import FlashAdamW

CKPT_DIR = "/tmp/flashoptim_dcp_repro"
MODEL = "Qwen/Qwen3-0.6B"
DATASET = "allenai/tulu-3-sft-mixture"
SEQ_LEN = 512
BATCH_SIZE = 8
STEPS = 20


def load_batches(tokenizer, num_batches, seq_len=SEQ_LEN, batch_size=BATCH_SIZE):
    ds = load_dataset(DATASET, split=f"train[:{num_batches * batch_size}]")
    batches, buffer = [], []
    for row in ds:
        text = tokenizer.apply_chat_template(row["messages"], tokenize=False, add_generation_prompt=False)
        enc = tokenizer(text, truncation=True, max_length=seq_len, padding="max_length", return_tensors="pt")
        input_ids = enc["input_ids"].squeeze(0)
        labels = input_ids.clone()
        labels[labels == tokenizer.pad_token_id] = -100
        buffer.append((input_ids, labels))
        if len(buffer) == batch_size:
            batches.append({
                "input_ids": torch.stack([b[0] for b in buffer]),
                "labels": torch.stack([b[1] for b in buffer]),
            })
            buffer = []
    return batches


def train(model, optimizer, batches, n, start=0):
    losses = []
    for i in range(n):
        batch = {k: v.to("cuda") for k, v in batches[start + i].items()}
        out = model(**batch)
        out.loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        losses.append(out.loss.item())
    return losses


def save_optim(model, optimizer, path):
    os.makedirs(path, exist_ok=True)
    osd = get_optimizer_state_dict(
        model, optimizer,
        options=StateDictOptions(flatten_optimizer_state_dict=True),
    )
    dcp.save({"optim": osd}, checkpoint_id=path)


def load_optim(model, optimizer, path):
    osd = get_optimizer_state_dict(
        model, optimizer,
        options=StateDictOptions(flatten_optimizer_state_dict=True),
    )
    sd = {"optim": osd}
    dcp.load(sd, checkpoint_id=path)
    set_optimizer_state_dict(
        model, optimizer,
        optim_state_dict=sd["optim"],
        options=StateDictOptions(flatten_optimizer_state_dict=True),
    )


def build(mesh):
    torch.manual_seed(42)
    m = AutoModelForCausalLM.from_pretrained(MODEL, torch_dtype=torch.bfloat16, trust_remote_code=True).to("cuda")
    fully_shard(m, mesh=mesh)
    o = FlashAdamW(
        [p for p in m.parameters() if p.requires_grad],
        lr=1e-4, master_weight_bits=24,
    )
    return m, o


def main():
    dist.init_process_group("nccl")
    rank = dist.get_rank()
    torch.cuda.set_device(rank)
    mesh = init_device_mesh("cuda", (dist.get_world_size(),))

    tokenizer = AutoTokenizer.from_pretrained(MODEL, trust_remote_code=True)
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token
    batches = load_batches(tokenizer, 2 * STEPS)
    if rank == 0:
        print(f"Loaded {len(batches)} batches from {DATASET} (seq_len={SEQ_LEN})")

    if rank == 0:
        shutil.rmtree(CKPT_DIR, ignore_errors=True)
    dist.barrier()

    # Phase 1: train + save
    model, optim = build(mesh)
    losses1 = train(model, optim, batches, STEPS, start=0)
    if rank == 0:
        print("Phase 1 - train")
        for i, l in enumerate(losses1, 1):
            print(f"  step {i}: loss={l:.4f}")

    save_optim(model, optim, CKPT_DIR)
    dist.barrier()

    # Phase 2: fresh model, load optimizer, continue
    model2, optim2 = build(mesh)
    with torch.no_grad():
        for p1, p2 in zip(model.parameters(), model2.parameters()):
            p2.copy_(p1)
    load_optim(model2, optim2, CKPT_DIR)

    losses_orig = train(model, optim, batches, STEPS, start=STEPS)
    losses_resumed = train(model2, optim2, batches, STEPS, start=STEPS)

    if rank == 0:
        print("\nPhase 2 - continued training")
        print(f"  {'step':<6} {'original':<12} {'resumed':<12} {'delta':<12}")
        for i, (a, b) in enumerate(zip(losses_orig, losses_resumed), STEPS + 1):
            d = abs(a - b)
            flag = " <<<" if d > 0.1 else ""
            print(f"  {i:<6} {a:<12.4f} {b:<12.4f} {d:<12.6f}{flag}")
        mx = max(abs(a - b) for a, b in zip(losses_orig, losses_resumed))
        print(f"\n  max delta: {mx:.6f}  {'PASS' if mx < 0.01 else 'FAIL'}")

    dist.barrier()
    if rank == 0:
        shutil.rmtree(CKPT_DIR, ignore_errors=True)
    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

### Run

```bash
uv pip install flashoptim datasets transformers
torchrun --nproc-per-node 2 test_flashoptim_dcp.py

# or
uvx --with torch,transformers,flashoptim,datasets torchrun --nproc-per-node 2 test_flashoptim_dcp.py
```

## Observed output

```
Loaded 40 batches from allenai/tulu-3-sft-mixture (seq_len=512)
Phase 1 - train
  step 1: loss=2.0970
  step 2: loss=2.8162
  step 3: loss=3.2537
  step 4: loss=3.4082
  step 5: loss=2.7170
  step 6: loss=2.6255
  step 7: loss=2.6646
  step 8: loss=2.7052
  step 9: loss=2.8671
  step 10: loss=2.2139
  step 11: loss=3.1165
  step 12: loss=2.3675
  step 13: loss=2.8046
  step 14: loss=2.3608
  step 15: loss=2.3194
  step 16: loss=3.0652
  step 17: loss=2.2868
  step 18: loss=2.5535
  step 19: loss=2.5261
  step 20: loss=2.2093

Phase 2 - continued training
  step   original     resumed      delta
  21     2.3693       2.3693       0.000000
  22     2.2105       1784.9491    1782.738574  <<<
  23     2.1088       582.9604     580.851678   <<<
  24     2.1477       499.5914     497.443670   <<<
  25     1.8827       473.6243     471.741557   <<<
  26     2.2578       406.0717     403.813811   <<<
  27     3.0443       375.1688     372.124433   <<<
  28     2.4717       255.6391     253.167317   <<<
  29     2.5028       189.1637     186.660840   <<<
  30     2.4536       164.2197     161.766093   <<<
  31     2.3710       132.6684     130.297375   <<<
  32     2.3078       117.0058     114.697976   <<<
  33     2.6075       108.6988     106.091331   <<<
  34     2.0626       86.3957      84.333113    <<<
  35     2.6188       91.1382      88.519408    <<<
  36     2.4408       78.1478      75.706956    <<<
  37     2.1939       72.3397      70.145797    <<<
  38     2.1458       67.8557      65.709879    <<<
  39     1.8790       64.7829      62.903853    <<<
  40     2.4964       68.0697      65.573332    <<<

  max delta: 1782.738574  FAIL
```

## Expected behavior

After loading optimizer state from a DCP checkpoint, continued training should produce losses nearly identical to the original (non-interrupted) training run. The delta between original and resumed losses should be close to zero (< 0.01) for all steps.

## Analysis

- **Step 21 (first forward pass after resume)**: delta = 0.000000 — model weights were copied correctly.
- **Step 22 (first optimizer.step() after resume)**: loss explodes from 2.21 to 1784.95 — the optimizer state was silently corrupted during the DCP save/load round-trip.
- The resumed loss slowly decays over subsequent steps but remains orders of magnitude higher than the original, indicating the model weights were severely damaged by the first corrupted update and the optimizer is slowly recovering.

**Comment by hemildesai** [comment]
I was able to fix this issue with the following patch but would be good to have official support.
```python
# ---------------------------------------------------------------------------
# Monkey-patch: make FlashAdamW state_dict() return DTensors for DCP compat
# ---------------------------------------------------------------------------
def apply_flashoptim_dtensor_patch():
    """Patch FlashOptimizer so state_dict() returns DTensors when params are DTensors.

    FlashAdamW stores optimizer state as local tensors (needed for Triton
    kernels). state_dict() materialises these to bf16 plain tensors. DCP then
    treats them as replicated data and scrambles shards across ranks on load.

    Fix: after materialisation, wrap each state tensor as a DTensor that
    mirrors the corresponding parameter's placement. On load, __setstate__
    unwraps DTensors back to local tensors before re-quantising.
    """
    from torch.distributed._tensor import DTensor

    from flashoptim.optimizers import FlashOptimizer

    _orig_state_dict = FlashOptimizer.state_dict
    _orig_setstate = FlashOptimizer.__setstate__

    def _patched_state_dict(self):
        sd = _orig_state_dict(self)

        # Collect DTensor params in param_groups order (matches state dict int keys)
        ordered_params = []
        for group in self.param_groups:
            for param in group["params"]:
                ordered_params.append(param)

        has_dtensors = any(isinstance(p, DTensor) for p in ordered_params)
        if not has_dtensors:
            return sd

        for pid, param_state in sd["state"].items():
            if pid >= len(ordered_params):
                continue
            orig_param = ordered_params[pid]
            if not isinstance(orig_param, DTensor):
                continue
            mesh = orig_param.device_mesh
            placements = orig_param.placements
            for key, val in param_state.items():
                if isinstance(val, torch.Tensor) and val.dim() > 0 and not isinstance(val, DTensor):
                    param_state[key] = DTensor.from_local(val, mesh, placements)

        return sd

    def _patched_setstate(self, state):
        # Unwrap DTensors back to local tensors before _load_state_for_param
        # processes them — Triton kernels need plain CUDA tensors.
        opt_state = state.get("state", {})
        for param_state in opt_state.values():
            if isinstance(param_state, dict):
                for key, val in param_state.items():
                    if isinstance(val, DTensor):
                        param_state[key] = val.to_local()
        _orig_setstate(self, state)

    FlashOptimizer.state_dict = _patched_state_dict
    FlashOptimizer.__setstate__ = _patched_setstate
```

**Comment by josejg** [comment]
Thank you for reporting the issue @hemildesai. Added a fix for this in the latest patch https://github.com/databricks/flashoptim/commit/f55294177b331b691673ae4127e936f4815e635f

`pip install --upgrade flashoptim` should pick up the change

If you are curious about the implementation, state_dict wrapping is [here](https://github.com/databricks/flashoptim/blob/main/flashoptim/optimizers.py#L618-L654),
and unwrapping is [here](https://github.com/databricks/flashoptim/blob/main/flashoptim/optimizers.py#L937-L944). Also, added test coverage [here](https://github.com/databricks/flashoptim/blob/main/test/test_fsdp2.py#L728-L824)

**Comment by hemildesai** [comment]
Awesome, thanks a lot for the quick fix. Looks like the fix should be general so I will test it out in combination with TP/EP/PP.
